File: parquet_eda.r
library(duckdb)
library(arrow)
library(dplyr)
# https://duckdb.org/2021/12/03/duck-arrow.html
# https://duckdb.org/docs/sql/aggregates
ds <- open_dataset("parquet/checkouts")

ds %>%
    head() %>%
    collect()

ds %>%
    filter(CheckoutYear == "2014") %>%
    to_duckdb() %>%
    group_by(ItemType) %>%
    summarise(
        Count = n(),
        date = min(CheckoutDateTime)) %>%
    collect() %>%
    data.frame()

ds %>%
    to_duckdb() %>%
    filter(CheckoutYear == "2015")

# https://ursalabs.org/arrow-r-nightly/articles/dataset.html
bucket <- "https://ursa-labs-taxi-data.s3.us-east-2.amazonaws.com"
for (year in 2018:2019) {
  if (year == 2019) {
    # We only have through June 2019 there
    months <- 1:6
  } else {
    months <- 1:12
  }
  for (month in sprintf("%02d", months)) {
    dir.create(file.path("nyc-taxi", year, month), recursive = TRUE)
    try(download.file(
      paste(bucket, year, month, "data.parquet", sep = "/"),
      file.path("nyc-taxi", year, month, "data.parquet"),
      mode = "wb"
    ), silent = TRUE)
  }
}

File: parquet_eda.py

# %%
#import sys
#!{sys.executable} -m pip install duckdb pyarrow

# Need pyarrow > 6.0.0
# import sys
# !{sys.executable} -m pip install --upgrade pyarrow

import sys
!{sys.executable} -m pip install --upgrade duckdb
# %% 
import pandas as pd
import numpy as np
import duckdb

import pyarrow.dataset as ds
import pyarrow.parquet as pq

# %%
checks = pq.read_table('parquet/checkouts')
collection = pq.read_table('parquet/collection')
# %%
print(checks.num_rows)
checks.slice(length=10).to_pandas()


# %%
checks.filter(('CheckoutYear', '=', '2014'))


# %%
# We transform the dataset DuckDB relation
ddb = duckdb.arrow(checks)
# %%
# https://duckdb.org/docs/sql/aggregates
(ddb
    .filter("CheckoutYear == 2014")
    .aggregate("ItemType, count(*) as count, min(CheckoutDateTime) as date, last(CheckoutDateTime) as lastdate, first(CheckoutDateTime) as firstdate", "ItemType")
    .arrow()
    .to_pandas())
# %%
# Reads Parquet File to an Arrow Table
arrow_table = pq.read_table('integers.parquet')

# Gets Database Connection
con = duckdb.connect()

(con.execute('''
        SELECT SUM(data)
        FROM arrow_table
        WHERE data > 50
    ''')
    .fetch_arrow_table()
    .to_pandas())

# %%
